app/models/voxelytics/VoxelyticsService.scala

```diff
@@ -0,0 +1,205 @@
+                                 lastModified: Instant)
+
+object ArtifactChecksumEntry {
+  implicit val jsonFormat: OFormat[ArtifactChecksumEntry] = Json.format[ArtifactChecksumEntry]
+}
+
+class VoxelyticsService @Inject()(voxelyticsDAO: VoxelyticsDAO)(implicit ec: ExecutionContext) extends FoxImplicits {
+
+  def checkAuth(runId: ObjectId, user: User): Fox[Unit] =
+    for {
+      runUserId <- voxelyticsDAO.getUserIdForRun(runId)
+    } yield bool2Fox(user.isAdmin || runUserId == user._id)
+
+  def checkAuthForWorkflowCreation(runName: String, user: User): Fox[Unit] =
+    for {
+      runUserId <- voxelyticsDAO.getUserIdForRunOpt(runName, user._organization)
+    } yield bool2Fox(user.isAdmin || runUserId.forall(_ == user._id))
+
+  def runPublicWrites(run: RunEntry, tasks: List[TaskRunEntry]): JsObject =
+    Json.toJson(run).as[JsObject] ++ Json.obj(
+      "tasks" -> tasks.map(Json.toJson(_))
+    )
+
+  private def artifactEntryPublicWrites(artifact: ArtifactEntry): (String, JsObject) =
+    (artifact.name,
+     artifact.metadata ++
+       Json.obj(
+         "artifactId" -> artifact.artifactId.id,
+         "taskId" -> artifact.taskId.id,
+         "name" -> artifact.name,
+         "path" -> artifact.path,
+         "fileSize" -> artifact.fileSize,
+         "inodeCount" -> artifact.inodeCount,
+         "version" -> artifact.version,
+         "taskName" -> artifact.taskName
+       ))
+
+  def artifactsPublicWrites(artifacts: List[ArtifactEntry]): JsObject = {
+    val artifactsByTask = artifacts.groupBy(_.taskName)
+    JsObject(artifactsByTask.map(artifactKV => {
+      val taskName = artifactKV._1
+      val artifacts = artifactKV._2
+      (taskName, JsObject(artifacts.map(artifactEntryPublicWrites)))
+    }))
+  }
+
+  def workflowConfigPublicWrites(workflowConfig: JsObject, tasks: List[TaskEntry]): JsObject =
+    workflowConfig ++ Json.obj("tasks" -> JsObject(tasks.map(t => (t.name, t.config ++ Json.obj("task" -> t.task)))))
+
+  def aggregateBeginEndTime(runs: List[RunEntry]): (VoxelyticsRunState, Instant, Option[Instant]) = {
+    val state = runs.maxBy(_.beginTime).state
+    val beginTime = runs.map(_.beginTime).min
+    val endTime = Try(runs.flatMap(_.endTime).max).toOption
+
+    (state, beginTime, endTime)
+  }
+
+  def combineTaskRuns(allTaskRuns: List[TaskRunEntry], mostRecentRunId: ObjectId): List[TaskRunEntry] =
+    allTaskRuns
+      .filter(task => task.runId == mostRecentRunId)
+      .map(task => {
+        val thisTaskRuns = allTaskRuns.filter(t => t.taskName == task.taskName).sortBy(_.beginTime)
+        val nonWaitingTaskRuns = thisTaskRuns.filter(t => {
+          t.state == VoxelyticsRunState.RUNNING || t.state == VoxelyticsRunState.COMPLETE || t.state == VoxelyticsRunState.FAILED || t.state == VoxelyticsRunState.CANCELLED
+        })
+        if (nonWaitingTaskRuns.nonEmpty) {
+          nonWaitingTaskRuns.head
+        } else {
+          thisTaskRuns.head
+        }
```

Comment:
```suggestion
        nonWaitingTaskRuns.headOption.getOrElse(thisTaskRuns.head)
```
(though It would of course be great if the second `.head` could also be replaced by headOption, but that may change the function signature)